### PR TITLE
Fix unstoppable kubewatcher

### PR DESCRIPTION
**Pull request by life1347**
This PR aims to fix the following issues
1. watchSubscription is unstoppable even the watch is removed by the user.
2. add more checks to make sure that watchSubscription is stopped by the user or the unknow issues

### CODE REVIEW

**Comment by soamvasani** [comment]
Would you mind explaining the bug a bit?  Makes it faster to review if I know what problem this is fixing. Thanks :)

**Comment by life1347** [comment]
ResultChan of watcher is closed if the kubwatcher is deleted, so watchSubscription should print out "stopped" message and stop.

```
for {
    for {
        ev, more := <-ws.kubeWatch.ResultChan()
        if !more {
            log.Println("Watch stopped", ws.Watch.Metadata.Name)
            break
        }
    ...
    if atomic.LoadInt32(ws.stopped) == 0 {
        err := ws.restartWatch()
        if err != nil {
            log.Panicf("Failed to restart watch: %v", err)
        }
    }
}
```

However, the detection of ResultChan close is inside a nested for loops in current implementation, and the break statement only breaks internal for loop. So watchSubscription will keep trying to wait for events but failed and print stopped message endlessly.

```
for {
    if ws.isStopped() {
        ws.kubeWatch.Stop()
        break
    }
    ev, more := <-ws.kubeWatch.ResultChan()
    if !more {
        if ws.isStopped() {
            // watch is removed by user.
            log.Println("Watch stopped", ws.Watch.Metadata.Name)
            return
        } else {
            // watch closed due to unknown issues, restart it.
            log.Println("Watch stopped unexpectedly", ws.Watch.Metadata.Name)
            err := ws.restartWatch()
            if err != nil {
                log.Panicf("Failed to restart watch: %v", err)
            }
            continue
        }
    }
}
```

This PR removes the outer for loop, and check the close of ResultChan is because of trigger deletion or caused by unknown issues. If a watcher is stopped by the user, then stop it. Otherwise, restart the watcher.

File: kubewatcher/kubewatcher.go [hunk 1]
```diff
@@ -258,50 +257,71 @@ func getResourceVersion(obj runtime.Object) (string, error) {
...
-				break
-			}
-			rv, err := getResourceVersion(ev.Object)
-			if err != nil {
-				log.Printf("Error getting resourceVersion from object: %v", err)
+				return
 			} else {
-				log.Printf("rv=%v", rv)
-				ws.lastResourceVersion = rv
+				// watch closed due to unknown issues, restart it.
```
Comment (soamvasani) [inline kubewatcher/kubewatcher.go]: Nitpick -- it's not an unknown issue, I just didn't know why when I first wrote this code :) The watch has a timeout, after which it returns and needs to be restarted. Would you mind changing this comment and log statement to something like "watch timed out, restarting"?

File: kubewatcher/kubewatcher.go [hunk 2]
```diff
@@ -258,50 +257,71 @@ func getResourceVersion(obj runtime.Object) (string, error) {
...
-				log.Printf("Failed to serialize object: %v", err)
-				// TODO send a POST request indicating error
-			}
+		if ev.Type == watch.Error {
+			e := apierrs.FromObject(ev.Object)
+			log.Println("Watch error, retrying in a second: %v", e)
+			// Start from the beginning to get around "too old resource version"
+			ws.lastResourceVersion = ""
+			time.Sleep(time.Second)
+			continue
```
Comment (soamvasani) [inline kubewatcher/kubewatcher.go]: Hmm, the old code restarts the watch when this happens, but the new code tries to read from the watch again. I think you need to call `ws.restartWatch()` here.

File: kubewatcher/kubewatcher.go [hunk 3]
```diff
@@ -258,50 +257,71 @@ func getResourceVersion(obj runtime.Object) (string, error) {
...
-				log.Panicf("Failed to restart watch: %v", err)
-			}
+
+		// Event and object type aren't in the serialized object
+		headers := map[string]string{
+			"Content-Type":             "application/json",
+			"X-Kubernetes-Event-Type":  string(ev.Type),
+			"X-Kubernetes-Object-Type": reflect.TypeOf(ev.Object).Elem().Name(),
 		}
+		// Event and object type aren't in the serialized object
```
Comment (soamvasani) [inline kubewatcher/kubewatcher.go]: duplicated comment

File: kubewatcher/kubewatcher.go [hunk 4]
```diff
@@ -258,50 +257,71 @@ func getResourceVersion(obj runtime.Object) (string, error) {
...
 	for {
-		for {
-			ev, more := <-ws.kubeWatch.ResultChan()
-			if !more {
+		// check watchSubscription is stopped or not before waiting for event
+		// comes from the kubeWatch.ResultChan(). This fix the edge case that
+		// new kubewatch is created in the restartWatch() while the old kubewatch
+		// is being used in watchSubscription.stop().
+		if ws.isStopped() {
+			ws.kubeWatch.Stop()
```
Comment (soamvasani) [inline kubewatcher/kubewatcher.go]: (It's not wrong to have this call here, I'm just trying to make sure I understand the code correctly.)

File: kubewatcher/kubewatcher.go [hunk 5]
```diff
@@ -258,50 +257,74 @@ func getResourceVersion(obj runtime.Object) (string, error) {
...
 	for {
-		for {
-			ev, more := <-ws.kubeWatch.ResultChan()
-			if !more {
+		// check watchSubscription is stopped or not before waiting for event
+		// comes from the kubeWatch.ResultChan(). This fix the edge case that
+		// new kubewatch is created in the restartWatch() while the old kubewatch
+		// is being used in watchSubscription.stop().
+		if ws.isStopped() {
+			ws.stop()
```
Comment (soamvasani) [inline kubewatcher/kubewatcher.go]: I don't understand this either. If `ws.isStopped()` is true why do you need to call `ws.stop()`?

File: kubewatcher/kubewatcher.go [hunk 6]
```diff
@@ -258,50 +257,74 @@ func getResourceVersion(obj runtime.Object) (string, error) {
...
 			if err != nil {
 				log.Panicf("Failed to restart watch: %v", err)
 			}
+			continue
+		}
+		rv, err := getResourceVersion(ev.Object)
+		if err != nil {
+			log.Printf("Error getting resourceVersion from object: %v", err)
+		} else {
+			log.Printf("rv=%v", rv)
```
Comment (soamvasani) [inline kubewatcher/kubewatcher.go]: could you remove this log, it's too verbose, and was just here for debugging

**Pull request merged**